### ISSUE TITLE
Tidy up of where files come from (outside of the container)

### ISSUE THREAD

**Issue by gilesknap** [epics-containers/ec-cli#21]
This issue is the 'outside of the container' equivalent of https://github.com/epics-containers/ibek/issues/114.

At present interactions with helm, kubectl and podman are handled in:

- `ec` command line tool from epics-containers-cli `python`
- CI in beamline/domain repos `github actions` + `bash`
- CI in generic IOC repos `github actions` + `bash`
- system tests (CI) in ibek-support `github actions` + `bash`

The following configuration is shared by some or all of the above:

- what cluster and namespace we are using
- beamline / domain name
- helm registry for helm charts
- container registry for images
- naming conventions for charts  images generic iocs and ioc instance
- file names and locations for IOC instances
- location of logging service such as graylog

Goals for this issue are:

- try to wrap up all of the configuration into `epics-containers-cli`
  - note this means loading python and calling `ec` in CI
- update ec to work with the new model of single devcontainer (removal of dev-e7)
- review and simplify the ec CLI
- review the singleton helmchart approach to IOC instance versions (see #20)
- add support for docker (instead of podman) and docker compose (instead of K8S) @coretl I'd like to have this for the Oxfordshire EPICS meeting - I think I can get it done in 1-2 days of effort and won't spend any longer on it if that's not the case)

**Comment by gilesknap** [comment]
proposal for `ec` namespace

- generic-ioc
  -  build: build a generic IOC (from outside of the devcontainer - used for debugging when devcontainer wont start)
  - debug-last: start a container at the last successful step at the last build
  - versions: list the available image registry versions of a GIOC
  - work-on: add some support modules to the 'work list' meaning they are copied to host and mounted into the container - alters the devcontainer.json to do this (devcontainer restart required)
  - versions: list the versions of this generic IOC (from helm or git depending on #20
  - NOTE most generic IOC related work is handled inside a devcontainer. 
- k8s 
  - deploy: helm installs a published taggged IOC instance (with options for debugging helm rendering)
  - deploy-work: helm installs from the filesystem with a work tagged version (with options for debugging helm rendering)
  - attach: get the iocsh prompt
  - stop: set replicas to 0
  - restart: delete the pod 
  - exec: run bash in the running container instance
  - logs: get log of current or prev execution of an IOC instance
  - resources: list PVS, Pods, Deployments, Replica Sets in the cluster for a IOC (or maybe for a beamline?)
- compose
  - same as k8s (to be run on the server hosting the compose)
- logs  
  - TODO
- ci
  -  TODO are there other things we want to do in ci - is it a bad idea to allow users to do them?  
  - e.g. publish container to registry
  - review what else
- ROOT namespace for very common commands
  - ps: list running IOCs (in K8S - in compose it would be nice to have this but how to go find all the servers?)
  -

**Comment by gilesknap** [comment]
OK I'm pretty happy with this now.

@coretl  @GDYendell  Please take a look let me know if you hate anything ...

generic IOC CI now [looks like this](https://github.com/epics-containers/ioc-template/blob/dev/.github/workflows/build.sh) with this [actions yaml](https://github.com/epics-containers/ioc-template/blob/dev/.github/workflows/buiild.yml)

These scripts work locally and in CI and `ec` worries about whether you have `docker` or `podman` plus other domain knowledge that used to get shared around all the CI scripts. The CI uses a `requirements.txt` to install `ec` so that dependabot can make PRs for us.

generic IOC [acceptance tests](https://github.com/epics-containers/ioc-template/blob/dev/tests/run-tests.sh)

Beamline CI just validates the IOC instances [like this](https://github.com/epics-containers/bl45p/blob/main/ci_verify.sh) The CI does not need to publish anything because `ec deploy bl45p-ea-ioc-01 2023.10.1` goes and pulls the assets needed to make the helm chart directly out of the repo and we no longer use helm registries at all. 

ALL of the files above are fully generic and would work for any generic IOC or beamline repo.

Here is the current ec command namespace

```$ ec --help
                                                                                     
 Usage: ec [OPTIONS] COMMAND [ARGS]...                                               
                                                                                     
 EPICS Containers assistant CLI                                                      
                                                                                     
╭─ Options ─────────────────────────────────────────────────────────────────────────╮
│ --version                           Log the version of ec and exit                │
│ --domain              -d      TEXT  beamline or accelerator domain to use         │
│                                     [default: bl45p]                              │
│ --org                 -o      TEXT  git remote organisation of domain repos       │
│                                     [default:                                     │
│                                     https://github.com/epics-containers]          │
│ --repo                -r      TEXT  beamline or accelerator domain repository of  │
│                                     ioc instances                                 │
│                                     [default:                                     │
│                                     https://github.com/epics-containers/bl45p]    │
│ --namespace           -n      TEXT  kubernetes namespace to use [default: bl45p]  │
│ --log-level                   TEXT  Log level (DEBUG, INFO, WARNING, ERROR,       │
│                                     CRITICAL)                                     │
│                                     [default: WARN]                               │
│ --install-completion                Install completion for the current shell.     │
│ --show-completion                   Show completion for the current shell, to     │
│                                     copy it or customize the installation.        │
│ --help                              Show this message and exit.                   │
╰───────────────────────────────────────────────────────────────────────────────────╯
╭─ Commands ────────────────────────────────────────────────────────────────────────╮
│ cluster   Commands communicating with the k8s cluster. See 'ec cluster --help     │
│ dev       Commands for building, debugging containers. See 'ec dev --help'        │
│ ioc       Commands for managing IOCs in the cluster. See 'ec ioc --help'          │
│ ps        List the IOCs running in the current domain                             │
╰───────────────────────────────────────────────────────────────────────────────────╯
$ ec ioc --help 
                                                                                     
 Usage: ec ioc [OPTIONS] COMMAND [ARGS]...                                           
                                                                                     
 Commands for managing IOCs in the cluster. See 'ec ioc --help'                      
                                                                                     
╭─ Options ─────────────────────────────────────────────────────────────────────────╮
│ --help          Show this message and exit.                                       │
╰───────────────────────────────────────────────────────────────────────────────────╯
╭─ Commands ────────────────────────────────────────────────────────────────────────╮
│ attach        Attach to the IOC shell of a live IOC                               │
│ delete        Remove an IOC helm deployment from the cluster                      │
│ deploy        Pull an IOC helm chart version from the domain repo and deploy it   │
│               to the cluster                                                      │
│ deploy-local  Deploy a local IOC helm chart directly to the cluster with dated    │
│               beta version                                                        │
│ exec          Execute a bash prompt in a live IOC's container                     │
│ instances     List all versions of the IOC available in the helm registry         │
│ log-history   Open historical logs for an IOC                                     │
│ logs          Show logs for current and previous instances of an IOC              │
│ restart       Restart an IOC                                                      │
│ start         Start an IOC                                                        │
│ stop          Stop an IOC                                                         │
│ template      print out the helm template generated from a local ioc instance     │
╰───────────────────────────────────────────────────────────────────────────────────╯

$ ec dev --help 
                                                                                     
 Usage: ec dev [OPTIONS] COMMAND [ARGS]...                                           
                                                                                     
 Commands for building, debugging containers. See 'ec dev --help'                    
                                                                                     
╭─ Options ─────────────────────────────────────────────────────────────────────────╮
│ --help          Show this message and exit.                                       │
╰───────────────────────────────────────────────────────────────────────────────────╯
╭─ Commands ────────────────────────────────────────────────────────────────────────╮
│ build         Build a generic IOC container locally from a container project.     │
│ debug-last    Launches a container with the most recent image build. Useful for   │
│               debugging failed builds - if the last build failed it will start    │
│               the container after the most recent successful build step.          │
│ exec          Execute a command inside a running local IOC container              │
│ launch        Launch an IOC instance using configuration from a domain repo. Or   │
│               by passing a generic IOC image ID. Can be used for local testing of │
│               IOC instances. You may find the devcontainer a more convenient way  │
│               to do this.                                                         │
│ launch-local  Launch a locally built generic IOC from the local cache with tag    │
│               "local". This can only be run after doing an "ec dev build" in the  │
│               same generic IOC project folder.                                    │
│ stop          Stop a running local IOC container                                  │
│ versions      List the available versions of the generic IOC container image in   │
│               the registry                                                        │
│ wait-pv       Execute a command inside a running local IOC container              │
╰───────────────────────────────────────────────────────────────────────────────────╯
$ ec cluster --help 
                                                                                     
 Usage: ec cluster [OPTIONS] COMMAND [ARGS]...                                       
                                                                                     
 Commands communicating with the k8s cluster. See 'ec cluster --help                 
                                                                                     
╭─ Options ─────────────────────────────────────────────────────────────────────────╮
│ --help          Show this message and exit.                                       │
╰───────────────────────────────────────────────────────────────────────────────────╯
╭─ Commands ────────────────────────────────────────────────────────────────────────╮
│ monitor       Monitor the status of IOCs in a domain                              │
│ resources     Output information about a domain's cluster resources               │
╰───────────────────────────────────────────────────────────────────────────────────╯

**Comment by gilesknap** [comment]
Its all good. Closing.
